interface/ceed-preconditioning.c

```diff
@@ -0,0 +1,1993 @@
+              // Diagonal Only
+              const CeedScalar qf_value =
+                assembled_qf_array[q*layout[0] + (((e_in*num_comp+c_out)*
+                                                   num_eval_mode_out+e_out)*num_comp+c_out)*layout[1] + e*layout[2]];
+              if (fabs(qf_value) > qf_value_bound)
+                for (CeedInt n=0; n<num_nodes; n++)
+                  elem_diag_array[(e*num_comp+c_out)*num_nodes+n] +=
+                    bt[q*num_nodes+n] * qf_value * b[q*num_nodes+n];
+            }
+      }
+    }
+  }
+  ierr = CeedVectorRestoreArray(elem_diag, &elem_diag_array);
+  CeedChkBackend(ierr);
+  ierr = CeedVectorRestoreArray(assembled_qf, &assembled_qf_array);
+  CeedChkBackend(ierr);
+
+  // Assemble local operator diagonal
+  ierr = CeedElemRestrictionApply(diag_rstr, CEED_TRANSPOSE, elem_diag,
+                                  assembled, request); CeedChkBackend(ierr);
+
+  // Cleanup
+  if (is_pointblock) {
+    ierr = CeedElemRestrictionDestroy(&diag_rstr); CeedChkBackend(ierr);
+  }
+  ierr = CeedVectorDestroy(&assembled_qf); CeedChkBackend(ierr);
+  ierr = CeedVectorDestroy(&elem_diag); CeedChkBackend(ierr);
+  ierr = CeedFree(&eval_mode_in); CeedChkBackend(ierr);
+  ierr = CeedFree(&eval_mode_out); CeedChkBackend(ierr);
+  ierr = CeedFree(&identity); CeedChkBackend(ierr);
+
+  return CEED_ERROR_SUCCESS;
+}
+
+/**
+  @brief Core logic for assembling composite operator diagonal
+
+  @param[in] op             CeedOperator to assemble point block diagonal
+  @param[in] request        Address of CeedRequest for non-blocking completion, else
+                            CEED_REQUEST_IMMEDIATE
+  @param[in] is_pointblock  Boolean flag to assemble diagonal or point block diagonal
+  @param[out] assembled     CeedVector to store assembled diagonal
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Developer
+**/
+static inline int CeedOperatorLinearAssembleAddDiagonalCompositeCore(
+  CeedOperator op, CeedRequest *request, const bool is_pointblock,
+  CeedVector assembled) {
+  int ierr;
+  CeedInt num_sub;
+  CeedOperator *suboperators;
+  ierr = CeedOperatorGetNumSub(op, &num_sub); CeedChkBackend(ierr);
+  ierr = CeedOperatorGetSubList(op, &suboperators); CeedChkBackend(ierr);
+  for (CeedInt i = 0; i < num_sub; i++) {
+    ierr = CeedOperatorAssembleAddDiagonalCore(suboperators[i], request,
+           is_pointblock, assembled); CeedChkBackend(ierr);
+  }
+  return CEED_ERROR_SUCCESS;
+}
+
+/**
+  @brief Build nonzero pattern for non-composite operator
+
+  Users should generally use CeedOperatorLinearAssembleSymbolic()
+
+  @param[in] op      CeedOperator to assemble nonzero pattern
+  @param[in] offset  Offset for number of entries
+  @param[out] rows   Row number for each entry
+  @param[out] cols   Column number for each entry
+
+  @return An error code: 0 - success, otherwise - failure
+
+  @ref Developer
+**/
+int CeedSingleOperatorAssembleSymbolic(CeedOperator op, CeedInt offset,
```

Comment:
`static` if they can't be called by a user. (I don't think these should be callable by a user.)

Reply:
Good call. I'll do a pass for that while I'm seeing if I can reactor any of the full assembly to work reduce overlap with the diagonal assembly.